### PR TITLE
Decouple the DaggerWorkerFactory using plugins

### PR DESCRIPTION
**Pull request by aitorvs**
Task/Issue URL: https://app.asana.com/0/0/1199633113249884/f
Tech Design URL: 
CC: 

**Description**:
Use dagger multibindings to decouple the `DaggerWorkerFactory`.

The `DaggerWorkerFactory` is a big object that creates the workers AND inject its dependencies. For that reason it needs to know and depend on ALL dependencies all the app workers require. Which makes it a god object.
In this PR I use dagger multibindings to setup a kind of plugin system where the `DaggerWorkerFactory` can delegate the injection of dependencies to the worker plugins.

**Steps to test this PR**:
1. install/update the app
2. app should work as before
3. all workes should just work as before

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

**Review by marcosholgado — CHANGES_REQUESTED** [review]
Nice job @aitorvs! I noticed the clear data notification does not show up anymore, left a couple of comments about that.

File: app/src/main/java/com/duckduckgo/app/notification/AndroidNotificationScheduler.kt [hunk 1]
```diff
@@ -107,3 +109,45 @@ class NotificationScheduler(
         const val UNUSED_APP_WORK_REQUEST_TAG = "com.duckduckgo.notification.schedule"
     }
 }
+
+class ClearDataNotificationWorkerInjectorPlugin(
+    private val notificationManagerCompat: NotificationManagerCompat,
+    private val notificationDao: NotificationDao,
+    private val notificationFactory: NotificationFactory,
+    private val pixel: Pixel,
+    private val privacyProtectionNotification: PrivacyProtectionNotification
```
[the remark this answers is not in the document]
Comment (aitorvs) [reply app/src/main/java/com/duckduckgo/app/notification/AndroidNotificationScheduler.kt:118]: good catch! done

File: app/src/main/java/com/duckduckgo/app/global/plugins/worker/WorkerPluginsModule.kt [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    ): WorkerInjectorPlugin = DataClearingWorkerInjectorPlugin(settingsDataStore, clearDataAction)
+
+    @Provides
+    @IntoSet
+    fun clearDataNotificationWorkerInjectorPlugin(
+        notificationManagerCompat: NotificationManagerCompat,
+        notificationDao: NotificationDao,
+        notificationFactory: NotificationFactory,
+        pixel: Pixel,
+        privacyProtectionNotification: PrivacyProtectionNotification
```
[the remark this answers is not in the document]
Comment (aitorvs) [reply app/src/main/java/com/duckduckgo/app/global/plugins/worker/WorkerPluginsModule.kt:54]: good catch! done

**Comment by aitorvs** [comment]
> Nice job @aitorvs! I noticed the clear data notification does not show up anymore, left a couple of comments about that.

Thanks @marcosholgado, good catch! it should be fixed now!

**Review by marcosholgado — APPROVED** [review]
Nice one! 👍

**Pull request merged**